[FIXED JENKINS-31185] Added hg support

matthewcmorgan opened the pull request:
First pass attempt to add hg support.  
Tested working on privately hosted Jenkins v. 1.635.  
Only supports web hook payload, not post service payload, though that wouldn't be hard to add.

Resolves [Jenkins-31185](https://issues.jenkins-ci.org/browse/JENKINS-31185).

Xarthisius commented:
@fbelzunc is there anything blocking this pull request?

fbelzunc commented:
Hello, sorry I didn't have time to review this. I will do tomorrow and provide my feedback.

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java:
```diff
@@ -54,7 +54,7 @@ public void triggerMatchingJobs(String user, String url, String scm) {
                         }
                     } else
-                        LOGGER.log(Level.FINE, "{0} hasn't BitBucketTrigger set", job.getName());
+                    LOGGER.log(Level.FINE, "{0} hasn't BitBucketTrigger set", job.getName());
```
This line shouldn't be touched as there is not any modification.

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java:
```diff
@@ -85,10 +84,32 @@ private boolean match(SCM scm, URIish url) {
+            try {
+                URI hgUri = new URI(((MercurialSCM) scm).getSource());
+                String remote = url.toString();
+                if (almostMatches(hgUri, remote))
```
Please, after an `if` open/close brackets `{`, `}` even if there is only one line.

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java:
```diff
@@ -1,9 +1,6 @@
 package com.cloudbees.jenkins.plugins;
-
```
You don't need to touch anything here.

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java:
```diff
@@ -1,9 +1,6 @@
 package com.cloudbees.jenkins.plugins;
-
 import java.util.logging.Logger;
-
```
You don't need to touch anything here.

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java:
```diff
@@ -1,9 +1,6 @@
 import java.util.logging.Logger;
-
 import javax.servlet.http.HttpServletRequest;
-
```
You don't need to touch anything here.

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java:
```diff
@@ -31,55 +28,26 @@ public void processPayload(JSONObject payload, HttpServletRequest request) {
-        String scm = repo.has("scm") ? repo.getString("scm") : "git";
+        if (payload.has("repository")) {
+            JSONObject repo = payload.getJSONObject("repository");
+            LOGGER.info("Received commit hook notification for "+repo);
```
Please, at least change:

```
LOGGER.info("Received commit hook notification for "+repo);
```

per

```
LOGGER.info("Received commit hook notification for "+ repo);
```
Although the right way is:

```
LOGGER.log(Level.INFO, "Received commit hook notification for {0}", repo);
```

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java:
```diff
@@ -31,55 +28,26 @@ public void processPayload(JSONObject payload, HttpServletRequest request) {
+
+            probe.triggerMatchingJobs(user, url, scm);
+        } else if (payload.has("scm")) {
+            LOGGER.info("Received commit hook notification for hg " + payload);
```
Same structure needs to be done here:

```
LOGGER.log(Level.INFO, "Received commit hook notification for hg {0}", payload);
```

fbelzunc commented on src/test/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessorTest.java:
```diff
@@ -43,9 +43,19 @@ public void testProcessWebhookPayload() {
+                    .element("href", url)));
+
         payloadProcessor.processPayload(payload, request);
-        
```
You don't need to change this.

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java:
```diff
@@ -31,55 +28,26 @@ public void processPayload(JSONObject payload, HttpServletRequest request) {
-    },
-    "user": "marcus"
-}
-*/
```
You don't need to change this.

fbelzunc commented:
:ant: You only need to do small modifications. 

I would also suggest you change the title to the pull request to *[FIXED JENKINS-31185] Added hg support*. I believe that if you don't do that then Jira will not auto-find the Jira issue related to this pull request.

Thanks for contributing to BitBucket plugin!

matthewcmorgan commented:
Sorry for the sloppy pull request, I will make these fixes and resubmit soon.  Thank you for taking the time to review.

iNikem commented:
@matthewcmorgan friendly ping :) Do you plan to finish this or we can take it over? We are quite interested in the Mercurial support and were thinking about implementing it ourselves.

matthewcmorgan commented:
I plan on cleaning this up over the holidays, should be resubmitted soon.  :)

matthewcmorgan commented:
Okay, resubmitted with the fixes.  One line I couldn't undo as it wasn't complaining that I removed a newline but that I removed unnecessary tabs on the line.  If it matters I suppose we can use the online github editor to put them back in.  Everything else was cleaned up.  Let me know if anything else is needed.  Thank you for the review @fbelzunc

fbelzunc commented on src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java:
```diff
@@ -81,14 +83,38 @@ private boolean match(SCM scm, URIish url) {
+            try {
+                URI hgUri = new URI(((MercurialSCM) scm).getSource());
+                String remote = url.toString();
+                if (almostMatches(hgUri, remote)) {
```
You should use [MercurialStatus.looselyMatches](https://github.com/jenkinsci/mercurial-plugin/blob/17a01f2097e3a90568b42705c9a0d223695bdef5/src/main/java/hudson/plugins/mercurial/MercurialStatus.java#L61) instead.

matthewcmorgan replied:
Yeah, I saw that when I was developing this, and I remember I couldn't use it.  I tried implementing it again and realized why I included the almostMatches method...  The 
[MercurialStatus.looselyMatches](https://github.com/jenkinsci/mercurial-plugin/blob/17a01f2097e3a90568b42705c9a0d223695bdef5/src/main/java/hudson/plugins/mercurial/MercurialStatus.java#L61) method is not public unfortunately.

fbelzunc replied:
You are right.

matthewcmorgan replied:
Sure thing!

fbelzunc commented:
Thank you!
